google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/DelayTransactionStartUntilFirstWriteMockServerTest.java

```diff
@@ -0,0 +1,575 @@
+  }
+
+  @Test
+  public void testEnable() {
+    String prefix = dialect == Dialect.POSTGRESQL ? "spanner." : "";
+    try (Connection connection = createConnection(";delayTransactionStartUntilFirstWrite=true")) {
+      assertTrue(connection.isDelayTransactionStartUntilFirstWrite());
+      try (ResultSet resultSet =
+          connection.executeQuery(
+              Statement.of(
+                  String.format(
+                      "show variable %sdelay_transaction_start_until_first_write", prefix)))) {
+        assertTrue(resultSet.next());
+        assertTrue(resultSet.getBoolean(0));
+        assertTrue(
+            resultSet.getBoolean(
+                prefix.toUpperCase() + "DELAY_TRANSACTION_START_UNTIL_FIRST_WRITE"));
+        assertFalse(resultSet.next());
+      }
+    }
+    try (Connection connection = createConnection()) {
+      assertFalse(connection.isDelayTransactionStartUntilFirstWrite());
+      connection.execute(
+          Statement.of(
+              String.format("set %sdelay_transaction_start_until_first_write=true", prefix)));
+      assertTrue(connection.isDelayTransactionStartUntilFirstWrite());
+    }
+  }
+
+  @Test
+  public void testDisable() {
+    String prefix = dialect == Dialect.POSTGRESQL ? "spanner." : "";
+    try (Connection connection = createConnection()) {
+      assertFalse(connection.isDelayTransactionStartUntilFirstWrite());
+      try (ResultSet resultSet =
+          connection.executeQuery(
+              Statement.of(
+                  String.format(
+                      "show variable %sdelay_transaction_start_until_first_write", prefix)))) {
+        assertTrue(resultSet.next());
+        assertFalse(resultSet.getBoolean(0));
+        assertFalse(
+            resultSet.getBoolean(
+                prefix.toUpperCase() + "DELAY_TRANSACTION_START_UNTIL_FIRST_WRITE"));
+        assertFalse(resultSet.next());
+      }
+    }
+    try (Connection connection = createConnection(";delayTransactionStartUntilFirstWrite=true")) {
+      assertTrue(connection.isDelayTransactionStartUntilFirstWrite());
+      connection.execute(
+          Statement.of(
+              String.format("set %sdelay_transaction_start_until_first_write=false", prefix)));
+      assertFalse(connection.isDelayTransactionStartUntilFirstWrite());
+    }
+  }
+
+  @Test
+  public void testDefaultUsesRealTransactions() {
+    try (Connection connection = createConnection()) {
+      executeRandomQuery(connection);
+      connection.commit();
+
+      assertEquals(1, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+      ExecuteSqlRequest request = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class).get(0);
+      assertTrue(request.hasTransaction());
+      assertTrue(request.getTransaction().hasBegin());
+      assertTrue(request.getTransaction().getBegin().hasReadWrite());
+      assertEquals(1, mockSpanner.countRequestsOfType(CommitRequest.class));
+    }
+  }
+
+  @Test
+  public void testSingleQuery() {
```

Comment:
nit: rename to something a bit more clear. It seems we test that we never commit nor rollback, because there are no writes (thus we are using a singleUse read only transaction?)

Reply:
I renamed the tests in this class to more clearly reflect that they test transactions with a single query, ... etc.